File: Algorithms Analysis/Labs/Lab 3 G1-4/HashingLab1A.py
import time

class HashTable:
    def __init__(self, size, mode="LP"):
        self.capacity = size
        self.mode = mode # "LP" for Linear Probing, "DH" for Double Hashing
        self.table = [None] * self.capacity

    def _hash1(self, key):
        # TODO: Implement primary hash function using the Division Method
        return 0

    def _hash2(self, key):
        # TODO: Implement secondary hash function for Double Hashing
        return 1

    def insert(self, key):
        # TODO: Implement insertion logic with proper collision resolution
        return True

    def search(self, key):
        probes = 0
        found = False
        
        # TODO: Implement search logic.
        # CRITICAL: Increment 'probes' every single time you look at a slot index.
        
        return found, probes


def generate_insertion_keys(count):
    return [i * 10 for i in range(1, count + 1)] # 10, 20, 30...

def generate_unsuccessful_keys(count):
    return [i * 10 + 5 for i in range(1, count + 1)] # 15, 25, 35...


def main():
    m = 1009 # Assigned prime table size
    num_test_keys = 200
    
    with open("data.txt", "w") as f:
        f.write("#Alpha\tTime_LP\tTime_DH\tProbes_LP\tProbes_DH\n")
        
        print("Starting empirical profiling for Flavor 1 (Unsuccessful Searches)...")
        
        for i in range(1, 10):
            alpha = i / 10.0
            num_elements = int(m * alpha)
            
            insertion_keys = generate_insertion_keys(num_elements)
            ghost_keys = generate_unsuccessful_keys(num_test_keys)
            
            # ------------------------------------------
            # PROFILE LINEAR PROBING
            # ------------------------------------------
            lp_table = HashTable(m, mode="LP")
            for key in insertion_keys:
                lp_table.insert(key)
                
            lp_probes = 0
            start_lp = time.perf_counter()
            for ghost_key in ghost_keys:
                _, p = lp_table.search(ghost_key)
                lp_probes += p
            stop_lp = time.perf_counter()
            duration_lp = (stop_lp - start_lp) * 1_000_000
            avg_probes_lp = lp_probes / num_test_keys
            
            # ------------------------------------------
            # PROFILE DOUBLE HASHING
            # ------------------------------------------
            dh_table = HashTable(m, mode="DH")
            for key in insertion_keys:
                dh_table.insert(key)
                
            dh_probes = 0
            start_dh = time.perf_counter()
            for ghost_key in ghost_keys:
                _, p = dh_table.search(ghost_key)
                dh_probes += p
            stop_dh = time.perf_counter()
            duration_dh = (stop_dh - start_dh) * 1_000_000
            avg_probes_dh = dh_probes / num_test_keys
            
            f.write(f"{alpha:.2f}\t{duration_lp:.2f}\t{duration_dh:.2f}\t{avg_probes_lp:.2f}\t\t{avg_probes_dh:.2f}\n")
            print(f"Finished processing Load Factor (\u03B1): {alpha:.1f}")

    print("Profiling completed successfully. Results recorded in data.txt")

if __name__ == "__main__":
    main()

File: Algorithms Analysis/Labs/Lab 3 G1-4/HashingLab1A.cpp
#include <iostream>
#include <vector>
#include <chrono>
#include <fstream>

using namespace std;
using namespace std::chrono;

enum Mode { LINEAR_PROBING, DOUBLE_HASHING };

class HashTable {
private:
    int capacity;
    Mode mode;
    vector<int> table;
    
    int hash1(int key) {
        // TODO: Implement primary hash function using the Division Method
        return 0; 
    }

    int hash2(int key) {
        // TODO: Implement secondary hash function for Double Hashing
        return 1; 
    }

public:
    HashTable(int size, Mode m) : capacity(size), mode(m) {
        table.resize(capacity, -1); // Initialize empty slots with -1
    }

    // Inserts a key into the table. Returns true if successful.
    bool insert(int key) {
        // TODO: Implement insertion logic with proper collision resolution.
        return true;
    }

    // Searches for a key. Returns a pair: <found (bool), probes_counted (int)>
    pair<bool, int> search(int key) {
        int probes = 0;
        bool found = false;
        // TODO: Implement search and prob count logic.
        
        
        return {found, probes};
    }
};

// Generates the crowded structured data (Arithmetic Progressions)
vector<int> generateInsertionKeys(int count) {
    vector<int> data;
    for (int i = 1; i <= count; i++) {
        data.push_back(i * 10); // 10, 20, 30, 40...
    }
    return data;
}

// Generates "Ghost" keys that are guaranteed NOT to be in the table
vector<int> generateUnsuccessfulKeys(int count) {
    vector<int> data;
    for (int i = 1; i <= count; i++) {
        data.push_back(i * 10 + 5); // 15, 25, 35, 45... (Guaranteed mismatch)
    }
    return data;
}

int main() {
    int m = 1009; // Assigned prime table size
    int numTestKeys = 200; // Number of lookup operations to average out timing
    
    ofstream outFile("data.txt");
    outFile << "#Alpha\tTime_LP\tTime_DH\tProbes_LP\tProbes_DH\n";

    cout << "Starting empirical profiling for Flavor 1 (Unsuccessful Searches)..." << endl;

    for (double alpha = 0.1; alpha <= 0.95; alpha += 0.1) {
        int numElements = static_cast<int>(m * alpha);
        
        vector<int> insertionKeys = generateInsertionKeys(numElements);
        vector<int> ghostKeys = generateUnsuccessfulKeys(numTestKeys);

        // ==========================================
        // PROFILE LINEAR PROBING
        // ==========================================
        HashTable lpTable(m, LINEAR_PROBING);
        for (int key : insertionKeys) {
            lpTable.insert(key); // Fill table statically to target alpha
        }

        long long lpTotalProbes = 0;
        auto startLP = high_resolution_clock::now();
        for (int ghostKey : ghostKeys) {
            auto result = lpTable.search(ghostKey);
            lpTotalProbes += result.second; // Accumulate unsuccessful probe steps
        }
        auto stopLP = high_resolution_clock::now();
        auto durationLP = duration_cast<microseconds>(stopLP - startLP).count();
        double avgProbesLP = (double)lpTotalProbes / numTestKeys;

        // ==========================================
        // PROFILE DOUBLE HASHING
        // ==========================================
        HashTable dhTable(m, DOUBLE_HASHING);
        for (int key : insertionKeys) {
            dhTable.insert(key); // Fill table statically to target alpha
        }

        long long dhTotalProbes = 0;
        auto startDH = high_resolution_clock::now();
        for (int ghostKey : ghostKeys) {
            auto result = dhTable.search(ghostKey);
            dhTotalProbes += result.second;
        }
        auto stopDH = high_resolution_clock::now();
        auto durationDH = duration_cast<microseconds>(stopDH - startDH).count();
        double avgProbesDH = (double)dhTotalProbes / numTestKeys;

        // Export data row
        outFile << alpha << "\t\t" << durationLP << "\t\t" << durationDH << "\t\t" 
                << avgProbesLP << "\t\t\t" << avgProbesDH << "\n";
                
        cout << "Finished processing Load Factor (\u03B1): " << alpha << endl;
    }

    outFile.close();
    cout << "Profiling completed successfully. Results recorded in data.txt" << endl;
    return 0;
}